Make breakpoints persistent across Zed sessions

Anthony-Eid opened the pull request:
Release Notes:

This PR aims to make breakpoint persistence across different sessions of zed & gets zed to send breakpoints from unopened files to debug adapters when necessary.

- [x] Serialize breakpoint to be able to save them in Zed's database
- [x] Save breakpoints in zed's database
- [x] Send breakpoints from unopened files to debug adapters when necessary
- [x] Handle breakpoint overlaps with code action button
- [x] Handle breakpoint overlaps with code runner button
- [x] Color line numbers debugger.accent color if they have a breakpoint on them
- [x] Add debugger.accent color to themes, allowing users to change breakpoint colors
- [x] Add 'save_breakpoints' setting
- [x] Fix bug where toggle breakpoint action wouldn't remove breakpoints added through clicking on the gutter
- [x] Send breakpoints from unopened files to debug bug adapters
- [ ] testing
- [x] documentation
- [x] Figure out how we want to handle a the case of a line deletion where the line directly correspond to a breakpoint
- [x] Enable rendering and toggling of breakpoints within a multi buffer


Note: While this PR is close to being completed I still want time to refine and clean up how I've serialized breakpoints within Zed. I also want to try a different approach to serializing breakpoints that should allow more flexibility in the future

Future Work:
- Change breakpoint rendering to create an element instead of using an icon (Would enable right click on breakpoints)
- Add different types of breakpoints (log, hit, conditional, etc)
- Fix bug where breakpoints don't show in a multi buffer if there isn't a singular buffer already open that contains the breakpoints
    - This will probably involve making buffer_store responsible for handing breakpoints state from open -> close & close_open
- Fix multi buffer bug where toggling a breakpoint within a multi buffer with an except_id other than 1 then opening the buffer associated with the breakpoint as a singleton causes zed to crash. This only happens when both the multi buffer and singleton buffer are both opened at the same time.
  
Current State:
- The below video shows breakpoint staying persistent between different instances of zed
- Line numbers with breakpoints on them are colored the breakpoint accent color
- Breakpoints handle overlaps with code action/runner symbols
- Breakpoints are visible within a multi buffer



https://github.com/user-attachments/assets/5c83b79b-2069-4ad7-85ae-b9a7cc4bce8b

Anthony-Eid committed:
Start work on setting up workplace serialization for breakpoints

Anthony-Eid committed:
Change breakpoints DS to use project path as key instead of bufferID

Anthony-Eid committed:
Get workplace to serialize breakpoints

Anthony-Eid committed:
Get load workspace to get breakpoint data

Anthony-Eid committed:
Change breakpoint DS back so it uses buffer_id as a key

I also made a new DS specifically for breakpoints that aren't part of any open buffers.
It should be easier to serialize and deserialize breakpoints now.

Anthony-Eid committed:
Start work on deserializing Workspace

Anthony-Eid committed:
Get deserialization to work for file's that open when initializing Zed

Anthony-Eid committed:
Transfer breakpoints from closed DS -> open DS when a buffer is open

Anthony-Eid committed:
Transfer breakpoints from open -> close when buffer maintaining them is released

Anthony-Eid committed:
Editor doesn't need to know about closed breakpoint

Anthony-Eid committed:
Introduce serialized breakpoint type & refactor code base to use it

Anthony-Eid committed:
Send breakpoints from unopened files to adapters & fix workspace serialization

Anthony-Eid committed:
Delete println that was used for debugging

RemcoSmitsDev commented on crates/dap/src/client.rs:
```diff
@@ -534,14 +534,14 @@ impl DebugAdapterClient {
     pub async fn set_breakpoints(
         &self,
-        path: PathBuf,
-        breakpoints: Option<Vec<SourceBreakpoint>>,
```
This should be still an option. That way you can remove them all from a file.

Anthony-Eid replied:
I'm open to changing it back, but I don't think that edge case is an issue. Whenever a breakpoint is toggled, updated breakpoint information is sent to active adapters. The only other case where breakpoints are sent to an adapter is during configuration which isn't affected by this change.

If we need to remove all breakpoints from a file we can send an empty Vec instead of None.

Anthony-Eid committed:
Fix bug where toggle breakpoint action wouldn't untoggle

Anthony-Eid committed:
Handle case where Code action indicator interlaps with breakpoint

Code action's that overlap with breakpoints will now show as red. Still need to
add a code action to toggle a breakpoint & handle overlaps with code runners

Anthony-Eid committed:
Show a breakpoint on line with code action while code action symbol is loading

I also handle overlaps with code test buttons as well

Anthony-Eid committed:
Get project::has_active_debuggers to return true if a debugger is running instead of starting

We shouldn't consider a debugger active until it's running because it may fail during it's starting phase.
Also, this check is used determine if we should update a file's breakpoints when one is toggled.

Anthony-Eid committed:
Fix breakpoint line number not being red when cursor is on it & clippy warnings

Anthony-Eid committed:
Get breakpoints to display in the correct positions within multi buffers

Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Merge branch 'remco-debugger' into breakpoints

Anthony-Eid committed:
Fix bug that caused a breakpoints to only remain persistent in one open file when booting zed up

Anthony-Eid committed:
Allow toggling breakpoints to work from within multibuffer gutter

Anthony-Eid committed:
Add breakpoint documentation

Anthony-Eid committed:
Add debugger.accent as a configurable theme

RemcoSmitsDev commented:
Would be nice if we have a settings to determine wether we persist the breakpoints.

RemcoSmitsDev commented on crates/editor/src/editor.rs:
```diff
@@ -5136,19 +5148,105 @@ impl Editor {
             .on_click(cx.listener(move |editor, _e, cx| {
                 editor.focus(cx);
-                editor.toggle_breakpoint_at_row(position, cx) //TODO handle folded
+                editor.toggle_breakpoint_at_anchor(position, cx) //TODO handle folded
```
I think this todo can be removed, we already handle this case:
https://github.com/RemcoSmitsDev/zed/pull/13/files#diff-2c4beb0f3febb4181a2f3bcf1d1008cd789d2f072b34b57c3de3249d88ffe48fR1590

Anthony-Eid replied:
I was thinking we could color a toggle the debugger accent color if there's a breakpoint nested within it. That's still not completed

RemcoSmitsDev commented on crates/workspace/src/persistence.rs:
```diff
@@ -419,6 +494,50 @@ impl WorkspaceDb {
             .flatten()?;
 
+        // dbg!
+        // Figure out why the below query didn't work
```
I think this debug code can be removed

RemcoSmitsDev commented on crates/workspace/src/persistence.rs:
```diff
@@ -709,6 +863,37 @@ impl WorkspaceDb {
         }
     }
 
+    // TODO: Fix this query
```
This is what you want to fix in a new PR right? So you can open all the breakpoints in a multi buffer?

Anthony-Eid committed:
Merge branch 'debugger' into breakpoints

RemcoSmitsDev commented on crates/workspace/Cargo.toml:
```diff
@@ -35,6 +35,8 @@ call.workspace = true
 client.workspace = true
 clock.workspace = true
 collections.workspace = true
+dap-types = { git = "https://github.com/zed-industries/dap-types" }
```
This can be removed right, because they are already inside the dap crate, and we export them there.

Anthony-Eid committed:
Breakpoint PR review edits

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Set up ground work for debugger settings & save breakpoints setting

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Merge branch 'debugger' into breakpoints

Anthony-Eid committed:
Add todo to fix bug & bug information

Co-authored-by: Remco Smits <djsmits12@gmail.com>

Anthony-Eid committed:
Merge debugger into breakpoints

Co-authored-by: Remco Smits <djsmits12@gmail.com>
